students/thecthaeh/Lesson05/mailroom_part3.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+
+"""Automate the process of creating 'Thank You' letters and tracking donors and donations through reports.
+
+"""
+import sys
+from operator import itemgetter
+
+#convert to a dict
+donor_info_dict = {}
+donor_info_dict['Frank Miller'] = [320.00, 100.00, 570.50]
+donor_info_dict['Tess Baker'] = [1000.00, 540.00]
+donor_info_dict['Grant Hugh'] = [5000.00]
+donor_info_dict['Sarah Piper'] = [40.00]
+donor_info_dict['Jim Newton'] = [1350.00, 1500.00, 5.50]
+
+def single_thank_you():
+    """Send a Thank You letter to a single donor."""
+    full_name = input("Enter the donor's full name.\n>> ")
+    
+    while full_name == 'list':
+        for donor in donor_info_dict:
+            print(donor)
+        full_name = input("\nEnter the donor's full name.\n>> ")
+        
+    if full_name.lower() == "quit":
+        return
+    
+    donate_amt = input("\nEnter the donation amount.\n>> $")
+    while donate_amt.lower() != 'quit':
+        try:
+            add_donation(full_name, float(donate_amt))
+            break
+        except ValueError:
+            donate_amt = input("\nInvalid entry, try again. \nUsing only numbers, enter the donation amount. \n>> $")
+    
+    if donate_amt.lower() == 'quit':
+        return
+
+def add_donation(donor_name, donate_amt):
+    """Add a donor and/or a donation."""
+    if donor_name in donor_info_dict:
+        donor_info_dict[donor_name].append(donate_amt)
+    else:
+        donor_info_dict.setdefault(donor_name, [donate_amt])
+        
+    print(f"Thank you, {donor_name}, for your generous donation of ${donate_amt:.2f}.")
+
+def create_report():
+    """Create a report listing donors, their total donation amount, the number of donations, and their average donation."""
+    header = ['Donor Name', 'Total Given', 'Num Gifts', 'Average Gifts']
+    
+    top_row = "{:30} |{:>20} |{:>15} |{:>20}".format(*header[:])
+    print(top_row)
+    print('-' * len(top_row))
+    
+    report = [[donor, sum(donor_info_dict[donor]), len(donor_info_dict[donor]), sum(donor_info_dict[donor])/len(donor_info_dict[donor])] for donor in donor_info_dict]
```

Comment:
you can make this line much shorter and more readable if you use items() (which is preferred anyway)
```
report = [[name, sum(donations), len(donations), sum(donations)/len(donations) for name, donations in donor_info_dict.items()]
```